### PR TITLE
IGNITE-19501 SchemaManager should use CatalogService for building BinaryRow descriptors

### PR DESCRIPTION
**Pull request by rpuch**
https://issues.apache.org/jira/browse/IGNITE-19501

### CODE REVIEW

File: modules/schema/src/test/java/org/apache/ignite/internal/schema/CatalogSchemaManagerTest.java [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+    void schemaRegistryByCausalityTokenIsAvailable() {
+        createSomeTable();
+
+        CompletableFuture<SchemaRegistry> future = schemaManager.schemaRegistry(CAUSALITY_TOKEN_1, TABLE_ID);
+        assertThat(future, willCompleteSuccessfully());
+
+        SchemaRegistry schemaRegistry = future.join();
+
+        assertThat(schemaRegistry.schema().version(), is(1));
+        assertThat(schemaRegistry.schema(1).version(), is(1));
```
[the remark this answers is not in the document]
Comment (rpuch) [reply modules/schema/src/test/java/org/apache/ignite/internal/schema/CatalogSchemaManagerTest.java:360]: I added such a test. Please take a look if this is what you wanted

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/CatalogSchemaManager.java [hunk 2]
```diff
@@ -0,0 +1,487 @@
...
+            Map<Integer, SchemaRegistryImpl> registries,
+            int tableId,
+            SchemaDescriptor schema
+    ) {
+        ByteArray key = schemaWithVerHistKey(tableId, schema.version());
+
+        byte[] serializedSchema = SchemaSerializerImpl.INSTANCE.serialize(schema);
+
+        return metastorageMgr.invoke(notExists(key), put(key, serializedSchema), noop())
+                .thenApply(t -> {
```
Comment (AMashenkov) [inline modules/schema/src/main/java/org/apache/ignite/internal/schema/CatalogSchemaManager.java:224]: Let's move this to the outside, and rename method to smth like `saveSchema`.
Thus, the method will represent a single action, its name will corresponds the content, and will `registries` changes will be localized.

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/CatalogSchemaManager.java [hunk 3]
```diff
@@ -0,0 +1,487 @@
...
+        return latestVersionFuture;
+    }
+
+    /**
+     * Gets the defined version of the table schema which available in Metastore.
+     *
+     * @param tblId Table id.
+     * @return Schema representation if schema found, {@code null} otherwise.
+     */
+    private CompletableFuture<SchemaDescriptor> schemaByVersion(int tblId, int ver) {
```
Comment (AMashenkov) [inline modules/schema/src/main/java/org/apache/ignite/internal/schema/CatalogSchemaManager.java:448]:
```suggestion
    private CompletableFuture<SchemaDescriptor> loadSchema(int tblId, int ver) {
```

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/CatalogSchemaManager.java [hunk 4]
```diff
@@ -0,0 +1,487 @@
...
+            }
+        });
+
+        return latestVersionFuture;
+    }
+
+    /**
+     * Gets the defined version of the table schema which available in Metastore.
+     *
+     * @param tblId Table id.
```
Comment (AMashenkov) [inline modules/schema/src/main/java/org/apache/ignite/internal/schema/CatalogSchemaManager.java:445]:
```suggestion
     * Loads the table schema descriptor by version from Metastore.
     *
     * @param tblId Table id.
     * @param ver Schema version.
```

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/CatalogSchemaManager.java [hunk 5]
```diff
@@ -0,0 +1,487 @@
...
+
+                SchemaDescriptor desc = registry0.schemaCached(schemaVer);
+
+                assert desc != null : "Unexpected empty schema description.";
+
+                fut.complete(desc);
+            }
+        };
+
+        registriesVv.whenComplete(schemaListener);
```
Comment (AMashenkov) [inline modules/schema/src/main/java/org/apache/ignite/internal/schema/CatalogSchemaManager.java:284]: The current method is called only from inside the SchemaRegistryImpl, and can be moved into the registry class. This listener can be replaced with VersionTracker to wait for the next versions (like CatalogService.versionTracker).

Thus, code will be much simple. RegistersVv calls will gone as registry is already available.
Joggling with subscriptions will be replaced with versionTracker future.

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/CatalogSchemaManager.java [hunk 6]
```diff
@@ -0,0 +1,487 @@
...
+    /**
+     * Gets the latest version of the table schema which is available in Metastore.
+     *
+     * @param tblId Table id.
+     * @return The latest schema version.
+     */
+    private CompletableFuture<Integer> latestSchemaVersion(int tblId) {
+        var latestVersionFuture = new CompletableFuture<Integer>();
+
+        metastorageMgr.prefix(schemaHistPrefix(tblId)).subscribe(new Subscriber<>() {
```
Comment (AMashenkov) [inline modules/schema/src/main/java/org/apache/ignite/internal/schema/CatalogSchemaManager.java:409]: I think we can update 2 keys in single transaction
* write a new descriptor to the history
* update latest version key with a version from written descriptor.

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/CatalogSchemaManager.java [hunk 7]
```diff
@@ -0,0 +1,487 @@
...
+        return registriesVv.latest().get(tableId);
+    }
+
+    /**
+     * Drop schema registry for the given table id.
+     *
+     * @param causalityToken Causality token.
+     * @param tableId Table id.
+     */
+    public CompletableFuture<?> dropRegistry(long causalityToken, int tableId) {
```
Comment (AMashenkov) [inline modules/schema/src/main/java/org/apache/ignite/internal/schema/CatalogSchemaManager.java:376]: Seems, we also should drop history from metastorage for the given table along with its registry.

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/CatalogSchemaManager.java (hunk 5 above)
Comment (rpuch) [reply modules/schema/src/main/java/org/apache/ignite/internal/schema/CatalogSchemaManager.java:315]: Fixed in https://issues.apache.org/jira/browse/IGNITE-20071

**Review by AMashenkov — APPROVED** [review]

**Pull request merged**